From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hash <38797169+BitBangingBytes@users.noreply.github.com>
Date: Thu, 19 Dec 2024 15:29:16 -0600
Subject: [PATCH] 7304: Fix M16C SFR memory size, add missing register.

Added missing PCR Special Function Register to M16C_60 and fixed SFR memory length in M16C_60 and M16C_80. It should be 0x400 not 0x3FF.
---
 Ghidra/Processors/M16C/data/languages/M16C_60.pspec | 5 +++--
 Ghidra/Processors/M16C/data/languages/M16C_80.pspec | 4 ++--
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/Ghidra/Processors/M16C/data/languages/M16C_60.pspec b/Ghidra/Processors/M16C/data/languages/M16C_60.pspec
index a1e280dda3..7434499fea 100644
--- a/Ghidra/Processors/M16C/data/languages/M16C_60.pspec
+++ b/Ghidra/Processors/M16C/data/languages/M16C_60.pspec
@@ -153,6 +153,7 @@
     <symbol name="PUR0" address="RAM:03FC"/>
     <symbol name="PUR1" address="RAM:03FD"/>
     <symbol name="PUR2" address="RAM:03FE"/>
+    <symbol name="PCR" address="RAM:03FF"/>
     <symbol name="UNDEFINED_INSTRUCTION_INT_VECTOR" address="RAM:FFFDC" entry="true" type="code_ptr"/>
     <symbol name="OVERFLOW_INT_VECTOR" address="RAM:FFFE0" entry="true" type="code_ptr"/>
     <symbol name="BRK_INSTRUCTION_INT_VECTOR" address="RAM:FFFE4" entry="true" type="code_ptr"/>
@@ -164,6 +165,6 @@
     <symbol name="RESET_INT_VECTOR" address="RAM:FFFFC" entry="true" type="code_ptr"/>
   </default_symbols>
   <default_memory_blocks>
-    <memory_block name="SFR" start_address="RAM:0000" mode="rw" length="0x03FF" initialized="false"/>
+    <memory_block name="SFR" start_address="RAM:0000" mode="rw" length="0x0400" initialized="false"/>
   </default_memory_blocks>
-</processor_spec>
\ No newline at end of file
+</processor_spec>
diff --git a/Ghidra/Processors/M16C/data/languages/M16C_80.pspec b/Ghidra/Processors/M16C/data/languages/M16C_80.pspec
index caf732d193..8538f2f5f7 100644
--- a/Ghidra/Processors/M16C/data/languages/M16C_80.pspec
+++ b/Ghidra/Processors/M16C/data/languages/M16C_80.pspec
@@ -291,6 +291,6 @@
     <symbol name="RESET_INT_VECTOR" address="RAM:FFFFFC" entry="true" type="code_ptr"/>
   </default_symbols>
   <default_memory_blocks>
-    <memory_block name="SFR" start_address="RAM:0000" mode="rw" length="0x03FF" initialized="false"/>
+    <memory_block name="SFR" start_address="RAM:0000" mode="rw" length="0x0400" initialized="false"/>
   </default_memory_blocks>
-</processor_spec>
\ No newline at end of file
+</processor_spec>
-- 
2.45.1

